### PR TITLE
Add avro_schema_nested field to schema_registry_decode / encode 

### PR DESCRIPTION
**Pull request by jem-davies**

### CODE REVIEW

File: internal/impl/confluent/processor_schema_registry_decode.go [hunk 1]
```diff
@@ -48,6 +48,9 @@ This processor decodes protobuf messages to JSON documents, you can read more ab
 		Field(service.NewBoolField("avro_raw_json").
 			Description("Whether Avro messages should be decoded into normal JSON (\"json that meets the expectations of regular internet json\") rather than [Avro JSON](https://avro.apache.org/docs/current/specification/_print/#json-encoding). If `true` the schema returned from the subject should be decoded as [standard json](https://pkg.go.dev/github.com/linkedin/goavro/v2#NewCodecForStandardJSONFull) instead of as [avro json](https://pkg.go.dev/github.com/linkedin/goavro/v2#NewCodec). There is a [comment in goavro](https://github.com/linkedin/goavro/blob/5ec5a5ee7ec82e16e6e2b438d610e1cab2588393/union.go#L224-L249), the [underlining library used for avro serialization](https://github.com/linkedin/goavro), that explains in more detail the difference between the standard json and avro json.").
 			Advanced().Default(false)).
+		Field(service.NewBoolField("avro_nested_schemas").
+			Description("Whether Avro Schemas are nested. If true bento will resolve schema references.").
+			Advanced().Default(false).Version("?.?.?")). // TODO: replace with the version numb
```
Comment (jem-davies) [inline internal/impl/confluent/processor_schema_registry_decode.go:53]: Will need to update this when it gets merged.

File: internal/impl/confluent/processor_schema_registry_encode.go [hunk 2]
```diff
@@ -70,7 +70,10 @@ We will be considering alternative approaches in future so please [get in touch]
...
 		Field(service.NewBoolField("avro_raw_json").
 			Description("Whether messages encoded in Avro format should be parsed as normal JSON (\"json that meets the expectations of regular internet json\") rather than [Avro JSON](https://avro.apache.org/docs/current/specification/_print/#json-encoding). If `true` the schema returned from the subject should be parsed as [standard json](https://pkg.go.dev/github.com/linkedin/goavro/v2#NewCodecForStandardJSONFull) instead of as [avro json](https://pkg.go.dev/github.com/linkedin/goavro/v2#NewCodec). There is a [comment in goavro](https://github.com/linkedin/goavro/blob/5ec5a5ee7ec82e16e6e2b438d610e1cab2588393/union.go#L224-L249), the [underlining library used for avro serialization](https://github.com/linkedin/goavro), that explains in more detail the difference between standard json and avro json.").
-			Advanced().Default(false).Version("1.0.0"))
+			Advanced().Default(false).Version("1.0.0")).
+		Field(service.NewBoolField("avro_nested_schemas").
+			Description("Whether Avro Schemas are nested. If true bento will resolve schema references.").
+			Advanced().Default(false).Version("?.?.?")) // TODO: replace with the version number
```
Comment (jem-davies) [inline internal/impl/confluent/processor_schema_registry_encode.go:76]: Will need to update this when it gets merged.

**Review by jem-davies — COMMENTED** [review]
reminder to update version numbers

File: internal/impl/confluent/processor_schema_registry_decode.go (hunk 1 above)
Comment (gregfurman) [reply internal/impl/confluent/processor_schema_registry_decode.go:53]: Let's make this `1.2.0` and try get it out in the next release!

Comment (jem-davies) [reply internal/impl/confluent/processor_schema_registry_decode.go]: added a test and changed the version numbers

File: internal/impl/confluent/serde_avro_test.go [hunk 3]
```diff
@@ -141,3 +141,99 @@ func TestAvroReferences(t *testing.T) {
...
 	}
 }
+
+func TestAvroReferencesNested(t *testing.T) {
+	tCtx, done := context.WithTimeout(context.Background(), time.Second*10)
+	defer done()
+
+	userSchema := `{"type":"record","name":"User","namespace":"com.example","fields":[{"name":"name","type":"string"},{"name":"email","type":"string"},{"name":"address","type":"Address"}]}`
+	addressSchema := `{"type":"record","name":"Address","namespace":"com.example","fields":[{"name":"street","type":"string"},{"name":"city","type":"string"},{"name":"state","type":"State"},{"name":"zip","type":"string"}]}`
+	stateSchema := `{"type":"record","name":"State","namespace":"com.example","fields":[{"name":"state","type":"string"}]}`
```
Comment (gregfurman) [inline internal/impl/confluent/serde_avro_test.go:151]: nit: can we format these? Would make reading the tests a bit easier.

File: internal/impl/confluent/serde_avro.go [hunk 4]
```diff
@@ -45,10 +58,72 @@ func resolveAvroReferences(ctx context.Context, client *schemaRegistryClient, in
 	return string(schemaHydratedBytes), nil
 }
 
-func (s *schemaRegistryEncoder) getAvroEncoder(ctx context.Context, info SchemaInfo) (schemaEncoder, error) {
-	schema, err := resolveAvroReferences(ctx, s.client, info)
+func resolveAvroReferencesNested(ctx context.Context, client *schemaRegistryClient, info SchemaInfo) (string, error) {
```
Comment (gregfurman) [inline internal/impl/confluent/serde_avro.go:61]: nit: Wow nice job here, man. Maybe some comments would be nice

File: internal/impl/confluent/serde_avro.go [hunk 5]
```diff
@@ -45,10 +58,72 @@ func resolveAvroReferences(ctx context.Context, client *schemaRegistryClient, in
...
+	var schema reference
+	err := json.Unmarshal(schemaRaw, &schema)
 	if err != nil {
-		return nil, err
+		return "", fmt.Errorf("failed to unmarshal root schema: %w", err)
+	}
+
+	var ref reference
+
+	for {
```
Comment (gregfurman) [inline internal/impl/confluent/serde_avro.go:90]: Just wanna clarify there isn't a case where this loop will never break?

**Review by gregfurman — COMMENTED** [review]
Ran all the tests locally and seems to work. If possible, maybe some more context on the solution would be great. Otherwise, 1-2 small nits and a question re: that `while` loop when resolving nested references.

File: internal/impl/confluent/serde_avro.go (hunk 5 above)
Comment (jem-davies) [reply internal/impl/confluent/serde_avro.go]: there was a possibility that this loop would never break, if there was a circular reference in the schemas, I have added a max limit to the loop, and add a unit test for this case.

File: internal/impl/confluent/serde_avro.go [hunk 6]
```diff
@@ -86,8 +86,12 @@ func resolveAvroReferencesNested(ctx context.Context, client *schemaRegistryClie
 	}
 
 	var ref reference
+	maxIterations := 100
 
-	for {
+	for i := 0; i <= maxIterations; i++ {
+		if i == maxIterations {
+			return "", fmt.Errorf("maximum iteration limit reached trying to resolve Avro references: possible circular dependency detected")
```
Comment (gregfurman) [inline internal/impl/confluent/serde_avro.go]: Nice! Maybe in future we can add something that'll check for these circular dependencies beforehand. For now, I think this is a fair fix. Wonder if it's worth mentioning in the field docs that we have a maximum depth of 100

File: internal/impl/confluent/processor_schema_registry_decode.go [hunk 7]
```diff
@@ -48,6 +48,9 @@ This processor decodes protobuf messages to JSON documents, you can read more ab
 		Field(service.NewBoolField("avro_raw_json").
 			Description("Whether Avro messages should be decoded into normal JSON (\"json that meets the expectations of regular internet json\") rather than [Avro JSON](https://avro.apache.org/docs/current/specification/_print/#json-encoding). If `true` the schema returned from the subject should be decoded as [standard json](https://pkg.go.dev/github.com/linkedin/goavro/v2#NewCodecForStandardJSONFull) instead of as [avro json](https://pkg.go.dev/github.com/linkedin/goavro/v2#NewCodec). There is a [comment in goavro](https://github.com/linkedin/goavro/blob/5ec5a5ee7ec82e16e6e2b438d610e1cab2588393/union.go#L224-L249), the [underlining library used for avro serialization](https://github.com/linkedin/goavro), that explains in more detail the difference between the standard json and avro json.").
 			Advanced().Default(false)).
+		Field(service.NewBoolField("avro_nested_schemas").
+			Description("Whether Avro Schemas are nested. If true bento will resolve schema references. (Up to a maximum depth of 100)").
+			Advanced().Default(false).Version("1.2.0")).
```
Comment (gregfurman) [inline internal/impl/confluent/processor_schema_registry_decode.go:53]:
```suggestion
			Description("Whether Avro Schemas are nested. If true, bento will resolve schema references. (Up to a maximum depth of 100)").
			Advanced().Default(false).Version("1.2.0")).
```

**Review by gregfurman — APPROVED** [review]
LGTM 🫡

**Pull request merged**